File: 剑指Offer04.二维数组中的查找/剑指Offer04.二维数组中的查找.py
'''
Descroption: 剑指 Offer 04. 二维数组中的查找
Author: EmoryHuang
Date: 2021-06-27 16:00:00
解题思路:
简单题，从左下或者右上进行查找
'''


class Solution:
    def findNumberIn2DArray(self, matrix: List[List[int]], target: int) -> bool:
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return False
        n, m = len(matrix), len(matrix[0])
        i, j = n-1, 0
        while i >= 0 and j < m:
            if matrix[i][j] == target:
                return True
            elif matrix[i][j] > target:
                i -= 1
            else:
                j += 1
        return False

File: Solution/剑指Offer04.二维数组中的查找/剑指Offer04.二维数组中的查找.cpp
class Solution {
public:
    bool findNumberIn2DArray(vector<vector<int>>& matrix, int target) {
        if (matrix.size() == 0) return false;
        int m = matrix.size();
        int n = matrix[0].size();
        int i = m - 1, j = 0;
        while (i >= 0 && j < n) {
            if (matrix[i][j] == target)
                return true;
            else if (matrix[i][j] > target)
                i--;
            else
                j++;
        }
        return false;
    }
};